file/file_util.cc

```diff
@@ -122,4 +122,53 @@ bool IsWalDirSameAsDBPath(const ImmutableDBOptions* db_options) {
   return same;
 }
 
+IOStatus GenerateOneFileChecksum(FileSystem* fs, const std::string& file_path,
+                                 FileChecksumGenFactory* checksum_factory,
+                                 std::string* file_checksum,
+                                 std::string* file_checksum_func_name) {
+  if (checksum_factory == nullptr) {
+    return IOStatus::InvalidArgument("Checksum factory is invalid");
+  }
+  assert(file_checksum != nullptr);
+  assert(file_checksum_func_name != nullptr);
+
+  FileChecksumGenContext gen_context;
+  std::unique_ptr<FileChecksumGenerator> checksum_generator =
+      checksum_factory->CreateFileChecksumGenerator(gen_context);
+  uint64_t size;
+  IOStatus io_s;
+  std::unique_ptr<SequentialFileReader> reader;
+  {
+    std::unique_ptr<FSSequentialFile> r_file;
+    io_s = fs->NewSequentialFile(file_path, FileOptions(), &r_file, nullptr);
+    if (!io_s.ok()) {
+      return io_s;
+    }
+    io_s = fs->GetFileSize(file_path, IOOptions(), &size, nullptr);
+    if (!io_s.ok()) {
+      return io_s;
+    }
+    reader.reset(new SequentialFileReader(std::move(r_file), file_path));
+  }
+
+  char buffer[4096];
+  Slice slice;
+  while (size > 0) {
+    size_t bytes_to_read = std::min(sizeof(buffer), static_cast<size_t>(size));
```

Comment:
> verify_checksums_readahead_size

This is a general function to generate the checksum of a file (can be any file in DB). So I think we do not need to use` verify_checksums_readahead_size` which is defined in IngestExternalFileOptions specifically.

Reply:
> Also, if size_t is 32 bits, bytes_to_read could be 0 (false report of Corruption) or very small (poor performance) for a ~4GB file.

You mean the issue caused by static_cast here? Or 4096 is too small?

Reply:
> You mean the issue caused by static_cast here? Or 4096 is too small?

static cast before min is a bug when size_t = 32 bits. I'm just noting that in some cases it could be a performance bug (and other cases correctness) as bytes_to_read could be some very small value like 1 for a very large file. 

> So I think we do not need to use verify_checksums_readahead_size which is defined in IngestExternalFileOptions specifically.

I would expect full file checksum should benefit more from readahead than block checksums, so it seems weird to me that verify_checksums_readahead_size would apply to verifying block checksums but not full file checksums, especially since the options are right next to each other in the source code.

I'm OK with "I'll fix that soon in another PR," but I don't want to go through a cycle of someone complaining about speed before we make it fast.

> This is a general function to generate the checksum of a file (can be any file in DB).

Perhaps add a parameter.

Reply:
@pdillinger Thanks for the explanation! Will add a readahead parameter to the interface. For the static cast, I will change to uint64_t size and cast sizeof(buffer) instead.